blueprint/core/blueprint_View.cpp

```diff
@@ -225,18 +252,93 @@ namespace blueprint
+        {
+            case juce::MouseInputSource::InputSourceType::mouse:
+                dispatchViewEvent("onMouseDown", detail::makeViewEventObject(e, *this));
+                break;
+            case juce::MouseInputSource::InputSourceType::touch: {
+                Touch* t = new Touch(e.source.getIndex(), e.x, e.y, e.getScreenX(), e.getScreenY(), getViewId());
+                touches.add(t);
+                targetTouches.add(t);
+
+                int indexInChangedTouches = getTouchIndexInList(e.source.getIndex(), changedTouches);
+                if (indexInChangedTouches == -1)
+                    changedTouches.add(t);
+
+                dispatchViewEvent("onTouchStart", detail::makeViewEventFromTouchList(touches, targetTouches, changedTouches, *this));
+                changedTouches.clear();
+                break;
+            }
+            case juce::MouseInputSource::InputSourceType::pen:
+                break;
+            default:
+                break;
+        }
+
+
     }
 
     void View::mouseUp (const juce::MouseEvent& e)
     {
-        dispatchViewEvent("onMouseUp", detail::makeViewEventObject(e, *this));
+        switch (e.source.getType())
+        {
+            case juce::MouseInputSource::InputSourceType::mouse:
+                dispatchViewEvent("onMouseUp", detail::makeViewEventObject(e, *this));
+                break;
+            case juce::MouseInputSource::InputSourceType::touch: {
+                Touch* t = touches.removeAndReturn(getTouchIndexInList(e.source.getIndex(), touches));
+
+                int indexInTargetTouches = getTouchIndexInList(e.source.getIndex(), targetTouches);
+                if (indexInTargetTouches != -1)
+                    targetTouches.remove(indexInTargetTouches);
+
+                int indexInChangedTouches = getTouchIndexInList(e.source.getIndex(), changedTouches);
+                if (indexInChangedTouches == -1)
+                    changedTouches.add(t);
+
+                dispatchViewEvent("onTouchEnd", detail::makeViewEventFromTouchList(touches, targetTouches, changedTouches, *this));
+                changedTouches.clear();
+                break;
+            }
+            case juce::MouseInputSource::InputSourceType::pen:
+                break;
+            default:
+                break;
+        }
     }
 
     void View::mouseDrag (const juce::MouseEvent& e)
     {
-        // TODO: mouseDrag isn't a dom event... is it?
-        dispatchViewEvent("onMouseDrag", detail::makeViewEventObject(e, *this));
+        switch (e.source.getType())
+        {
+            case juce::MouseInputSource::InputSourceType::mouse:
+                // TODO: mouseDrag isn't a dom event... is it?
+                dispatchViewEvent("onMouseDrag", detail::makeViewEventObject(e, *this));
+                break;
+            case juce::MouseInputSource::InputSourceType::touch: {
+                Touch* t = getTouch(e.source.getIndex(), touches);
```

Comment:
From a style point of view some white space between lines here would be nice to assist readability.